src/next_array.rs

```diff
@@ -0,0 +1,163 @@
+use core::mem::{self, MaybeUninit};
+use core::ptr;
+
+/// An array of at most `N` elements.
+struct ArrayBuilder<T, const N: usize> {
+    /// The (possibly uninitialized) elements of the `ArrayBuilder`.
+    ///
+    /// # Safety
+    ///
+    /// The elements of `arr[..len]` are valid `T`s.
+    arr: [MaybeUninit<T>; N],
+
+    /// The number of leading elements of `arr` that are valid `T`s, len <= N.
+    len: usize,
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    /// Initializes a new, empty `ArrayBuilder`.
+    pub fn new() -> Self {
+        // SAFETY: The safety invariant of `arr` trivially holds for `len = 0`.
+        Self {
+            arr: [(); N].map(|_| MaybeUninit::uninit()),
+            len: 0,
+        }
+    }
+
+    /// Pushes `value` onto the end of the array.
+    ///
+    /// # Panics
+    ///
+    /// This panics if `self.len >= N` or if `self.len == usize::MAX`.
+    pub fn push(&mut self, value: T) {
+        // PANICS: This will panic if `self.len >= N`.
+        // SAFETY: The safety invariant of `self.arr` applies to elements at
+        // indices `0..self.len` — not to the element at `self.len`. Writing to
+        // the element at index `self.len` therefore does not violate the safety
+        // invariant of `self.arr`. Even if this line panics, we have not
+        // created any intermediate invalid state.
+        self.arr[self.len] = MaybeUninit::new(value);
+        // PANICS: This will panic if `self.len == usize::MAX`.
+        // SAFETY: By invariant on `self.arr`, all elements at indicies
+        // `0..self.len` are valid. Due to the above write, the element at
+        // `self.len` is now also valid. Consequently, all elements at indicies
+        // `0..(self.len + 1)` are valid, and `self.len` can be safely
+        // incremented without violating `self.arr`'s invariant. It is fine if
+        // this increment panics, as we have not created any intermediate
+        // invalid state.
+        self.len = match self.len.checked_add(1) {
+            Some(sum) => sum,
+            None => panic!("`self.len == usize::MAX`; cannot increment `len`"),
+        };
```

Comment:
Isn't this panic branch unreachable? Since `self.arr[self.len]` did not panic above, we therefore know that `self.len < N`, meaning that `self.len + 1 <= N` and thus `self.len + 1` cannot overflow.
```suggestion
        // PANICS: This cannot panic, since `self.len < N <= usize::MAX`.
        // SAFETY: By invariant on `self.arr`, all elements at indicies
        // `0..self.len` are valid. Due to the above write, the element at
        // `self.len` is now also valid. Consequently, all elements at indicies
        // `0..(self.len + 1)` are valid, and `self.len` can be safely
        // incremented without violating `self.arr`'s invariant.
        self.len += 1;
```